lang/csharp/src/apache/main/POCO/POCOWriter.cs

```diff
@@ -0,0 +1,250 @@
+        /// <param name="schema"></param>
+        public POCODefaultWriter(Type objType, Schema schema)
+            : base(schema)
+        {
+            var rs = schema as RecordSchema;
+            if (rs != null)
+            {
+                ClassCache.LoadClassCache(objType, rs);
+            }
+        }
+
+        /// <summary>
+        /// Serialized a record using the given RecordSchema. It uses GetField method
+        /// to extract the field value from the given object.
+        /// </summary>
+        /// <param name="schema">The RecordSchema to use for serialization</param>
+        /// <param name="value">The value to be serialized</param>
+        /// <param name="encoder">The Encoder for serialization</param>
+
+        protected override void WriteRecord(RecordSchema schema, object value, Encoder encoder)
+        {
+
+            foreach (Field field in schema)
+            {
+                try
+                {
+                    var v = ClassCache.GetClass(schema).GetValue(value, field);
+
+                    Write(field.Schema, v, encoder);
+                }
+                catch (Exception ex)
+                {
+                    throw new AvroException(ex.Message + " in field " + field.Name, ex);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Validates that the record is a fixed record object and that the schema in the object is the
+        /// same as the given writer schema. Writes the given fixed record into the given encoder
+        /// </summary>
+        /// <param name="schema">writer schema</param>
+        /// <param name="value">fixed object to write</param>
+        /// <param name="encoder">encoder to write to</param>
+        protected override void WriteFixed(FixedSchema schema, object value, Encoder encoder)
+        {
+            var fixedrec = value as byte[];
+            if (fixedrec == null)
+            {
+                throw new AvroTypeException("Fixed object is not derived from byte[]");
+            }
+
+            if (fixedrec.Length != schema.Size)
+            {
+                throw new AvroTypeException($"Fixed object length is not the same as schema length {schema.Size}");
+            }
+
+            encoder.WriteFixed(fixedrec);
+        }
+
+        /// <summary>
+        /// Serialized an array. The default implementation calls EnsureArrayObject() to ascertain that the
+        /// given value is an array. It then calls GetArrayLength() and GetArrayElement()
+        /// to access the members of the array and then serialize them.
+        /// </summary>
+        /// <param name="schema">The ArraySchema for serialization</param>
+        /// <param name="value">The value being serialized</param>
+        /// <param name="encoder">The encoder for serialization</param>
+        protected override void WriteArray(ArraySchema schema, object value, Encoder encoder)
+        {
+            var arr = value as System.Collections.IList;
```

Comment:
Could we allow `IEnumerable<T>` here rather than requiring an `IList`? `DefaultSpecificWriter` also requires an `IList` at this time, but maybe we could imporve on this here. Would allowing an `IEnumerable` cause problems?